core/primitives/src/sr25519.rs

```diff
@@ -172,12 +222,87 @@ impl ::std::fmt::Debug for Public {
 	}
 }
 
+/// A since derivation junction description. It is the single parameter used when creating
+/// a new secret key from an existing secret key and, in the case of `SoftRaw` and `SoftIndex`
+/// a new public key from an existing public key.
+pub enum DeriveJunction {
+	/// Soft (vanilla) derivation. Public keys have a correspondent derivation.
+	Soft(ChainCode),
+	/// Hard ("hardened") derivation. Public keys do not have a correspondent derivation.
+	Hard(ChainCode),
+}
+
+impl DeriveJunction {
+	/// Consume self to return a soft derive junction with the same chain code.
+	pub fn soften(self) -> Self { DeriveJunction::Soft(self.unwrap_chain_code()) }
+	/// Consume self to return a hard derive junction with the same chain code.
+	pub fn harden(self) -> Self { DeriveJunction::Hard(self.unwrap_chain_code()) }
+
+	/// Consume self to return the chain code.
+	pub fn unwrap_chain_code(self) -> ChainCode {
+		match self {
+			DeriveJunction::Hard(c) | DeriveJunction::Soft(c) => c,
+		}
+	}
+	/// Return a reference to the chain code.
+	pub fn chain_code(&self) -> &ChainCode {
+		match *self {
+			DeriveJunction::Hard(ref c) | DeriveJunction::Soft(ref c) => c,
+		}
+	}
+	/// Return `true` if the junction is soft.
+	pub fn is_soft(&self) -> bool {
+		match *self {
+			DeriveJunction::Soft(_) => true,
+			_ => false,
+		}
+	}
+	/// Return `true` if the junction is hard.
+	pub fn is_hard(&self) -> bool {
+		match *self {
+			DeriveJunction::Hard(_) => true,
+			_ => false,
+		}
+	}
+/*
+	/// Derive a new key from the given key following this junction.
+	pub fn derive<Key: Derivable>(&self, key: &Key) -> Option<Key> {
+		match *self {
+			DeriveJunction::Soft(ref cc) => key.derive_soft()
+			DeriveJunction::Hard(ref cc) => key.derive_soft()
+		}
+	}*/
+}
+
+impl<T: Encode> From<T> for DeriveJunction {
+	fn from(index: T) -> Self {
```

Comment:
Wouldn't it be better to have an explicit method for that?, it would help with documenting the hashing behaviour and that we always end up with `Soft` junction. Such widely applied `From` impl might be a bit too magical imho

Reply:
Moved it to a normal `impl` function `soft`.